src/mpi/coll/algorithms/recexchalgo/recexchalgo.c

```diff
@@ -87,57 +95,64 @@ int MPII_Recexchalgo_get_neighbors(int rank, int nranks, int *k_,
      * The non-participating ranks send their data to the participating rank
      * in their corresponding set.
      */
-    T = (rem * k) / (k - 1);
-    *T_ = T;
-    *p_of_k_ = p_of_k;
+    rem = nranks - p_of_k;
+    int a = rem / (k - 1);
+    int b = rem % (k - 1);
+    T = a * k + b;
 
```

Comment:
I can't follow the details of the changes here. Can you explain if T changed from original version?

Reply:
Yes. 
```
T = (rem * k ) / (k - 1)   # substitute rem = a * (k - 1) + b
  = (a * (k-1) + b) * k  / (k -1)
  = a * (k-1) * k / (k-1) + b * k / (k-1)
  = a * k + b * (k-1 + 1) / (k-1)
  = a * k + b + b/(k-1)                      # since b = rem % (k-1), b / (k-1) < 1, so it gets rounded off.
  = a * k + b
```
`a * k + b` is easier to understand, since there are `a` groups of `k` procs, plus the leftover `b`.

Reply:
Is this a must-have change? Or can this formula be in the comments for easy understanding?
To calculate a and b, it takes an extra MOD operation which was not there.

Reply:
I don't understand the math of `T = rem * k / (k -1)`. If you can explain it so other developers can understand it, then please.

Reply:
We prefer our original code, but I also appreciate the ease understanding of yours. So I proposed to add your formula in the comments. No?

Reply:
As I said, I don't insist on one code vs another. But I insist on the code to be understandable and maintainable. I don't understand the original code, thus it is unmaintainable to me. If you can help me understand it, please do. Otherwise, it has no reason to stay. You are certainly allowed prefer it, but surely you have a reason for the preference? Please explain it so we can prefer it as well.

Reply:
You just explained it in your first comment, do I understand it correctly? You showed that T = (rem * k) / (k - 1) = a * k + b. What else need to be explained? 
I prefer the original code because it is concise and more efficient. The new formula has more calculation steps and extra "%" operation.

Reply:
My explanation is showing that the original code is a "crypt" version of an understandable code. The explanation does not justify that we should keep the "crypt" version.

Reply:
> I prefer the original code because it is concise and more efficient. The new formula has more calculation steps and extra "%" operation.

What's important is the steps a developer need to take to understand the code, which is many and non-obvious as in https://github.com/pmodels/mpich/pull/5620#discussion_r747768587

Reply:
I don't know if I made myself clear - I welcomed the explanation in the comments, but I prefer the original code. They are *equivalent* as you proved in https://github.com/pmodels/mpich/pull/5620#discussion_r747768587. The original code is concise and more efficient than yours.

Reply:
I disagree with your preference.

Reply:
Also, as I read the code, I don't find it too difficult to understand:
T = (rem * k ) / (k - 1)  = rem * (k/k-1)
What it means is that I want to expand non-participating ranks (rem) to form a group of k ranks.  In each group, the (k-1) is from the rem (non-participating ranks), the other one rank is the participating ranks. Therefore, I am combining the non-participating ranks and participating ranks together, which technically expanding rem by a ratio of k/(k-1). Therefore the formula.

Reply:
Okay, I'll put your explanation in. The explanation is all I ask for.

Reply:
Another maybe better way to understand it is that assuming T is the combined ranks, then we have:
T / n = rem / (n-1), because the number of groups are the same, so we get:
T = rem * n / (n-1)

Reply:
That doesn't explain the rounding.  `T/k = rem/(k-1) --> T = rem/(k-1)*k`, not `T = rem * k / (k-1)`.

Reply:
Updated with your explanation and added a note on how the extra processes worked with rounding.

Reply:
> That doesn't explain the rounding. `T/k = rem/(k-1) --> T = rem/(k-1)*k`, not `T = rem * k / (k-1)`.

ok, to consider rounding, it should be ceiling(T/K) == ceiling(rem/(k-1)). It probably leads to the same result after dividing into different condition of whether k-1 can divide rem. But I agree the explanation can be tedious, just that it may help to understand it. Let's not waste more time on it.

Reply:
> Let's not waste more time on it.

This is not a waste of time. In computer science, we can't even off by 1. We have to understand how the extra process worked. `T-1`, or `T-2`, ..., all the way to `T - T%k` all end up with the same number of groups, so you cannot explain `T` by just using the preservation of groups.